### PR TITLE
[lang] [refactor] Move logical operations to ops.py

### PR DESCRIPTION
**Pull request by archibate**
Related issue = https://github.com/taichi-dev/taichi/issues/1050#issuecomment-633396937

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: taichi/python/export_lang.cpp [hunk 1]
```diff
@@ -385,9 +385,8 @@ void export_lang(py::module &m) {
   m.def("expr_bit_and", expr_bit_and);
   m.def("expr_bit_or", expr_bit_or);
   m.def("expr_bit_xor", expr_bit_xor);
-  m.def("expr_bit_not", [](const Expr &expr) {
-    return Expr::make<UnaryOpExpression>(UnaryOpType::logic_not, expr);
-  });
```
Comment (archibate) [inline taichi/python/export_lang.cpp:390]: Broken window?

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 2]
```diff
@@ -657,7 +657,7 @@ class KernelGen : public IRVisitor {
   }
 
   void visit(IfStmt *if_stmt) override {
-    emit("if ({} != 0) {{", if_stmt->cond->short_name());
+    emit("if (({} & 1) != 0) {{", if_stmt->cond->short_name());
```
Comment (archibate) [inline taichi/backends/opengl/codegen_opengl.cpp:660]: Currently LLVM backends use -1 for true, while OpenGL backend uses 1 for true, so in afraid of misusing bit_not as logic_not, I have to do this for now.

**Comment by KLozes** [comment]
LGTM. even chained comparison worked! We should probably add some tests to test_element_wise too, though!

**Comment by archibate** [comment]
Will do in this PR. And I got
```
tests/python/test_element_wise.py::test_matrix_element_wise_unary_infix python3: /home/bate/Downloads/qq-files/1931127624/file_recv/llvm-8.0.1.src/lib/IR/Instructions.cpp:2156: void llvm::BinaryOperator::AssertOK(): Assertion `getType()->isIntOrIntVectorTy() && "Tried to create a logical operation on a non-integral type!"' failed.
```
when testing bit_not, while running correct in a separate file.

**Comment by archibate** [comment]
Sleeping now, feel free to merge this if you approve these changes @yuanming-hu! I will refactor `augassign`/`atomic_xxx` to ops.py iapr too.

File: python/taichi/misc/test.py [hunk 3]
```diff
@@ -1,10 +1,24 @@
...
-def approx(*args, **kwargs):
+class boolean_integer:
+    def __init__(self, value):
+        self.value = value
+
+    def __eq__(self, other):
+        if self.value:
+            return bool(other)
+        else:
+            return not bool(other)
```
Comment (yuanming-hu) [inline python/taichi/misc/test.py:12]:
```suggestion
        return bool(self.value) == bool(other)
```

File: taichi/python/export_lang.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/python/export_lang.cpp:390]: Nice catch!

**Comment by archibate** [comment]
Btw, I found the link in PR template:
http://taichi.readthedocs.io/en/latest/contributor_guide.html#pr-title-tags
is broken due to rename, it should be:
https://taichi.readthedocs.io/en/latest/contributor_guide.html#pr-title-format-and-tags
or
https://taichi.readthedocs.io/en/latest/contributor_guide.html#prtag
May I fix it itpr?

File: docs/syntax.rst [hunk 4]
```diff
@@ -165,14 +165,20 @@ Supported scalar functions:
...
+    def inside_taichi_scope():
+        x = 233
+        print('hello', x)
+        #=> hello 233
+
+        m = ti.Matrix([[2, 3, 4], [5, 6, 7]])
+        print('m is', m)
+        #=> m is [[2, 3, 4], [5, 6, 7]]
+
+        v ti.Vector([3, 4])
```
Comment (yuanming-hu) [inline docs/syntax.rst:182]:
```suggestion
        v = ti.Vector([3, 4])
```

**Comment by KLozes** [comment]
lets also move the casting operators to common ops as well!

https://github.com/taichi-dev/taichi/blob/0e10326a5f88bf4f916460559737e7975ad46999/python/taichi/lang/expr.py#L272-L278

**Comment by KLozes** [comment]
oops too late!

**Pull request merged**